File: app/schemas/saved_job.py
"""定义收藏职位及其分析上下文在 API、领域服务和 JSON 快照之间共用的 Pydantic 契约。"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

SavedJobStatus = Literal[
    "saved",
    "interested",
    "applied",
    "interviewing",
    "rejected",
    "closed",
    "archived",
]


class SavedJobAnalysis(BaseModel):
    saved_job_analysis_id: str
    saved_job_id: str
    user_id: str
    resume_profile_id: str | None = None
    source_job_search_run_id: str | None = None
    source_job_result_id: str | None = None
    match_score: int | None = None
    confidence_label: str | None = None
    recommendation: str | None = None
    matched_strengths: list[str] = Field(default_factory=list)
    critical_gaps: list[str] = Field(default_factory=list)
    resume_actions: list[str] = Field(default_factory=list)
    interview_questions: list[str] = Field(default_factory=list)
    analysis: dict[str, object] = Field(default_factory=dict)
    analysis_mode: str
    created_at: datetime


class SavedJobOrigin(BaseModel):
    saved_job_origin_id: str
    user_id: str
    saved_job_id: str
    origin_type: Literal["search_result", "browser_capture", "manual"]
    resume_profile_id: str | None = None
    job_search_run_id: str | None = None
    job_search_result_id: str | None = None
    saved_job_analysis_id: str | None = None
    profile_label: str | None = None
    search_query: str | None = None
    source_provider: str | None = None
    created_at: datetime


class SavedJob(BaseModel):
    saved_job_id: str
    user_id: str
    source_provider: str | None = None
    source_url: str | None = None
    normalized_source_key: str | None = None
    title: str
    company: str | None = None
    location: str | None = None
    salary: str | None = None
    employment_type: str | None = None
    raw_jd_text: str
    structured_jd: dict[str, object] = Field(default_factory=dict)
    tags: list[str] = Field(default_factory=list)
    status: SavedJobStatus = "saved"
    notes: str | None = None
    first_seen_at: datetime
    saved_at: datetime
    updated_at: datetime
    archived_at: datetime | None = None
    latest_analysis: SavedJobAnalysis | None = None


class SavedJobCreateRequest(BaseModel):
    """描述收藏职位create的输入结构。"""
    source_provider: str | None = None
    source_url: str | None = None
    title: str
    company: str | None = None
    location: str | None = None
    salary: str | None = None
    employment_type: str | None = None
    raw_jd_text: str
    structured_jd: dict[str, object] = Field(default_factory=dict)
    tags: list[str] = Field(default_factory=list)
    status: SavedJobStatus = "saved"
    notes: str | None = None

    @field_validator("title", "raw_jd_text")
    @classmethod
    def _required_string(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("field cannot be empty")
        return cleaned


class SavedJobUpdateRequest(BaseModel):
    """描述收藏职位update的输入结构。"""
    status: SavedJobStatus | None = None
    notes: str | None = None
    tags: list[str] | None = None


class SavedJobFromSearchResultRequest(BaseModel):
    """描述收藏职位from搜索结果的输入结构。"""
    job_search_run_id: str
    job_result_id: str
    resume_profile_id: str | None = None
    tags: list[str] = Field(default_factory=list)
    status: SavedJobStatus = "saved"
    notes: str | None = None


class SavedJobFromBrowserCaptureRequest(SavedJobCreateRequest):
    """描述收藏职位from浏览器职位采集的输入结构。"""
    analysis: dict[str, object] | None = None
    resume_profile_id: str | None = None
    match_score: int | None = None
    confidence_label: str | None = None
    recommendation: str | None = None


class SavedJobListResponse(BaseModel):
    items: list[SavedJob] = Field(default_factory=list)


class SavedJobAnalysisListResponse(BaseModel):
    items: list[SavedJobAnalysis] = Field(default_factory=list)


class SavedJobOriginListResponse(BaseModel):
    items: list[SavedJobOrigin] = Field(default_factory=list)


class SavedJobStatusEvent(BaseModel):
    saved_job_status_event_id: str
    saved_job_id: str
    user_id: str
    from_status: SavedJobStatus | None = None
    to_status: SavedJobStatus
    reason: str | None = None
    changed_at: datetime


class SavedJobStatusEventListResponse(BaseModel):
    items: list[SavedJobStatusEvent] = Field(default_factory=list)

File: web/src/types/savedJob.ts
export interface SavedJobAnalysis {
  saved_job_analysis_id: string;
  saved_job_id: string;
  user_id: string;
  resume_profile_id: string | null;
  source_job_search_run_id: string | null;
  source_job_result_id: string | null;
  match_score: number | null;
  confidence_label: string | null;
  recommendation: string | null;
  matched_strengths: string[];
  critical_gaps: string[];
  resume_actions: string[];
  interview_questions: string[];
  analysis: Record<string, unknown>;
  analysis_mode: string;
  created_at: string;
}

export interface SavedJob {
  saved_job_id: string;
  user_id: string;
  source_provider: string | null;
  platform_job_id: string | null;
  source_url: string | null;
  normalized_source_key: string | null;
  title: string;
  company: string | null;
  location: string | null;
  salary: string | null;
  employment_type: string | null;
  raw_jd_text: string;
  structured_jd: Record<string, unknown>;
  tags: string[];
  notes: string | null;
  first_seen_at: string;
  saved_at: string;
  updated_at: string;
  archived_at: string | null;
  latest_analysis: SavedJobAnalysis | null;
}

export interface SavedJobListResponse {
  items: SavedJob[];
}

export interface SavedJobAnalysisListResponse {
  items: SavedJobAnalysis[];
}

export interface JobBriefContent {
  decision_summary: string;
  fit_signals: string[];
  evidence_gaps: string[];
  resume_actions: string[];
  interview_focus: string[];
  next_actions: string[];
}

export interface JobBrief {
  job_brief_id: string;
  saved_job_id: string;
  user_id: string;
  resume_profile_id: string | null;
  source_analysis_id: string | null;
  version: number;
  content: JobBriefContent;
  analysis_mode: string;
  analysis_provider: string | null;
  fallback_reason: string | null;
  created_at: string;
}

export interface JobBriefListResponse {
  items: JobBrief[];
}

export interface JobBriefGeneratePayload {
  resume_profile_id?: string | null;
  llm_provider?: string | null;
}

export interface PreparationSkillGap {
  skill: string;
  importance: "high" | "medium" | "low";
  evidence_status: "supported" | "partial" | "unknown" | "missing";
  skill_type: "knowledge" | "experience";
  jd_evidence: string;
  profile_evidence: string[];
  rationale: string;
  evidence_origin: "resume" | "user_reported" | "none";
  dimensions: Array<{
    dimension_id: string;
    label: string;
    state: "unresolved" | "supported" | "demonstrated" | "partial" | "knowledge_gap" | "missing" | "unknown";
    evidence: string[];
  }>;
}

export type PreparationExperienceLevel =
  | "work_experience"
  | "project_experience"
  | "practice_only"
  | "conceptual_only"
  | "no_experience"
  | "uncertain";

export interface PreparationAnswerOption {
  option_id: string;
  answer_kind?: "evidence_claim" | "partial_practice" | "knowledge_gap" | "explicit_absence" | "unclear" | null;
  value: PreparationExperienceLevel;
  label: string;
  description: string;
  evidence_transition: "supported" | "partial" | "unknown" | "missing";
  route: "ask_evidence" | "learning" | "capability_gap" | "clarify" | "next_skill";
  detail_policy: "required" | "optional" | "not_needed";
  follow_up_prompt: string | null;
  decision_dimension: string;
  state_effects: Array<{
    dimension_id: string;
    state: "unresolved" | "supported" | "demonstrated" | "partial" | "knowledge_gap" | "missing" | "unknown";
  }>;
  next_question_signal: string;
}

export interface PreparationQuestion {
  question_id: string;
  skill: string;
  prompt: string;
  why_asked: string;
  options: PreparationAnswerOption[];
  free_text_allowed: boolean;
  free_text_prompt: string;
  decision_objective: {
    dimension_id: string;
    uncertainty: string;
    why_now: string;
  } | null;
}

export interface PreparationAnswer {
  question_id: string;
  response_mode: "option" | "free_text";
  selected_option_id?: string | null;
  free_text?: string | null;
  experience_level?: PreparationExperienceLevel | null;
  detail?: string | null;
  detail_quality?: "not_provided" | "specific" | "vague";
  evidence_transition?: "supported" | "partial" | "unknown" | "missing" | null;
  route?: "ask_evidence" | "learning" | "capability_gap" | "clarify" | "next_skill" | null;
  resolution_source?: "option" | "llm_classified" | "fallback_uncertain" | "legacy" | null;
  input_mode?: "option_only" | "option_with_detail" | "free_text" | null;
  follow_up_count?: number;
  pending_prompt?: string | null;
  committed?: boolean;
}

export interface LearningResource {
  topic: string;
  title: string;
  url: string;
  source: string;
  level: string;
  reason: string;
}

export interface PreparationRecommendation {
  title: string;
  action: string;
  action_type: "learning" | "experience_inventory" | "interview_story" | "capability_gap";
  skill: string | null;
  evidence_basis: string[];
  resource_urls: string[];
}

export interface PreparationGenerationStage {
  mode: "deterministic" | "llm" | "fallback";
  provider: string | null;
  prompt_version: string;
  attempts: number;
  fallback_reason: string | null;
  attempt_errors: string[];
}

export interface InterviewPreparationWorkspace {
  preparation_id: string;
  saved_job_id: string;
  user_id: string;
  resume_profile_id: string | null;
  source_analysis_id: string | null;
  status: "questions_ready" | "paused" | "completed" | "stopped";
  skill_gaps: PreparationSkillGap[];
  questions: PreparationQuestion[];
  answers: PreparationAnswer[];
  learning_resources: LearningResource[];
  recommendations: PreparationRecommendation[];
  analysis_mode: string;
  analysis_provider: string | null;
  fallback_reason: string | null;
  question_generation: PreparationGenerationStage | null;
  recommendation_generation: PreparationGenerationStage | null;
  resource_mode: string;
  resource_warning: string | null;
  created_at: string;
  updated_at: string;
}

export type ApplicationStage =
  | "not_started"
  | "contacted"
  | "recruiter_replied"
  | "resume_requested"
  | "resume_ready"
  | "resume_sent"
  | "interview"
  | "closed";

export type ApplicationNextAction =
  | "generate_greeting"
  | "review_greeting"
  | "wait_for_reply"
  | "review_reply"
  | "generate_resume"
  | "review_resume"
  | "send_resume"
  | "prepare_interview"
  | "none";

export interface JobApplication {
  application_id: string;
  user_id: string;
  saved_job_id: string;
  stage: ApplicationStage;
  next_action: ApplicationNextAction;
  last_activity_at: string;
  contacted_at: string | null;
  replied_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ApplicationEvent {
  event_id: string;
  application_id: string;
  user_id: string;
  event_type: string;
  source: "web" | "browser_helper" | "system" | "user";
  detail: string | null;
  metadata: Record<string, unknown>;
  created_at: string;
}

export interface CommunicationDraft {
  draft_id: string;
  application_id: string | null;
  browser_capture_id: string | null;
  generated_content: string;
  approved_content: string | null;
  status: "generated" | "approved" | "sent" | "failed" | "dismissed";
  evidence_used: string[];
  avoid_claims: string[];
  analysis_provider: string | null;
  created_at: string;
  updated_at: string;
  sent_at: string | null;
}

export interface TailoredResumeVersion {
  tailored_resume_id: string;
  user_id: string;
  saved_job_id: string;
  resume_profile_id: string;
  version: number;
  content: string;
  validation: {
    is_valid: boolean;
    issues: string[];
    warnings: string[];
  };
  status: "needs_review" | "approved";
  analysis_provider: string | null;
  created_at: string;
  updated_at: string;
  approved_at: string | null;
}

export interface SavedJobWorkspace {
  job: SavedJob;
  application: JobApplication | null;
  latest_analysis: SavedJobAnalysis | null;
  communication_draft: CommunicationDraft | null;
  tailored_resume: TailoredResumeVersion | null;
  allowed_stage_transitions: ApplicationStage[];
  events: ApplicationEvent[];
}

export interface SavedJobCreatePayload {
  source_provider?: string | null;
  platform_job_id?: string | null;
  source_url?: string | null;
  title: string;
  company?: string | null;
  location?: string | null;
  salary?: string | null;
  employment_type?: string | null;
  raw_jd_text: string;
  structured_jd?: Record<string, unknown>;
  tags?: string[];
  notes?: string | null;
}

export interface SavedJobUpdatePayload {
  notes?: string | null;
  tags?: string[] | null;
}

export interface SavedJobFromSearchResultPayload {
  job_search_run_id: string;
  job_result_id: string;
  resume_profile_id?: string | null;
  tags?: string[];
  notes?: string | null;
}
